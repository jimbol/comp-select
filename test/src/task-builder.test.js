const createTaskBuilders  = require('../../src/task-builder');
const { SELECTOR, STATIC, POPULATE } = require('../../src/constants');

const transformers = {
  filter: {
    type: SELECTOR,
    fn: (task) => (args) => {}
  },
  getEach: {
    type: STATIC,
    fn: (task) => (prevResult) => {}
  },
  populate: {
    type: POPULATE,
    fn: (task) => (prevResult) => {}
  },
};

describe('createTaskBuilders', () => {
  it('creates task builders for passed transformers', () => {
    const results = createTaskBuilders(transformers);
    const context = { tasks: [] };
    expect(typeof results.filter).toBe('function')
    expect(typeof results.getEach).toBe('function')
  });

  it('adds tasks to context when invoked', () => {
    const results = createTaskBuilders(transformers);
    const context = { tasks: [] };
    function fake(){}

    const fakeSelector = jest.fn();
    results.filter.apply(context, ['fakeResultFunc']);
    results.filter.apply(context, ['fakeSelector1', 'fakeSelector2', 'fakeResultFunc']);
    results.getEach.apply(context, ['fake.path', 'fakeFallbackValue']);
    results.populate.apply(context, [fake]);
    results.populate.apply(context, ['fake.path', fake]);
    results.populate.apply(context, [{ 'fake.path': fake }]);
    expect(context).toMatchSnapshot();
  });
});
