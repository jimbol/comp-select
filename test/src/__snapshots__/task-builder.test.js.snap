// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTaskBuilders adds tasks to context when invoked 1`] = `
Object {
  "tasks": Array [
    Object {
      "name": "filter",
      "resultFunc": "fakeResultFunc",
      "selectors": Array [],
    },
    Object {
      "name": "filter",
      "resultFunc": "fakeResultFunc",
      "selectors": Array [
        "fakeSelector1",
        "fakeSelector2",
      ],
    },
    Object {
      "name": "getEach",
      "staticArgs": Array [
        "fake.path",
        "fakeFallbackValue",
      ],
    },
    Object {
      "name": "populate",
      "selectors": Array [
        [Function],
      ],
    },
    Object {
      "name": "populate",
      "paths": Array [
        "fake.path",
      ],
      "selectors": Array [
        [Function],
      ],
    },
    Object {
      "name": "populate",
      "paths": Array [
        "fake.path",
      ],
      "selectors": Array [
        Array [
          "fake.path",
        ],
      ],
    },
  ],
}
`;
